sdk/webpubsub/azure-messaging-webpubsubclient/azure/messaging/webpubsubclient/_client.py

```diff
@@ -558,40 +693,50 @@ def _retry(self, func: Callable[[], None]):
                 )
                 time.sleep(delay_seconds)
 
-    def _call_back(self, callback_type: Union[CallbackType, str], *args):
+    def _call_back(self, callback_type: CallbackType, *args):
         for func in self._handler[callback_type]:
             # _call_back works in listener thread which must not be blocked so we have to execute the func
             # in new thread to avoid dead lock
             threading.Thread(target=func, args=args, daemon=True).start()
 
-    def _start_from_restarting(self):
+    def _start_from_restarting(self, reconnect_tried_times: Optional[int] = None):
         if self._state != WebPubSubClientState.DISCONNECTED:
             _LOGGER.warning("Client can be only restarted when it's Disconnected")
             return
 
         try:
-            self._start_core()
+            self._start_core(reconnect_tried_times)
         except Exception as e:
             self._state = WebPubSubClientState.DISCONNECTED
             raise e
 
-    def _handle_auto_reconnect(self):
+    def _handle_auto_reconnect(self, reconnect_tried_times: Optional[int] = None):
         _LOGGER.debug("start auto reconnect")
         success = False
-        attempt = 0
+        attempt = -1 if reconnect_tried_times is None else reconnect_tried_times
         while not self._is_stopping:
             try:
-                self._start_from_restarting()
+                attempt = attempt + 1
+                if attempt > 0:
```

Comment:
So it looks like if `reconnect_tried_times=None`, we retry without delay? That seems odd?

Reply:
For sync API, there is backgroud thread to receive message from service; for async, there is a similar task. (let's call them A). When connection is broken, A will receive close message from service and call back `on_close` registerd by client SDK and `on_close` will try to reconnect/recover. once `on_close` complete, A will be over. So we have to make sure another backgroud thread/task (let's call it AA) can be created successfully and work well which means reconnect/recover succeed before A done.
When AA is created but still receive close message, then AA will try to repeat the up process. So if client confirm AA is trying to create another thread/task (AAA) to go on reconnect/recover, client has to stop reconnect/recover of A([here](https://github.com/Azure/azure-sdk-for-python/blob/4f5a6b561862abc367f4c0096984033bc06866ac/sdk/webpubsub/azure-messaging-webpubsubclient/azure/messaging/webpubsubclient/_client.py#L735-L737)) to avoid duplicated reconnect/recover logic. It 
To avoid reconnect/recover persist indefinitely, before A done, A must pass AA the reconnected times already tried (`reconnect_tried_times`) so that AA or possible AA...A will not recover/reconnect without control.
Above all, if `reconnect_tried_times/recover_start_time` is None, it means the first time to reconnect/recover; if specific value, it means already tried in "pararent" thread/task and client shall continue with these values.